File: python/101-150/133. Clone Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: UndirectedGraphNode
        :rtype: UndirectedGraphNode
        """
        if not node:
            return None
        newnode = UndirectedGraphNode(node.label)
        dic = {}
        dic[node] = newnode
        queue = [node]
        while queue:
            cur = queue.pop(0)
            for n in cur.neighbors:
                if n in dic:
                    dic[cur].neighbors += [dic[n]]
                else:
                    temp = UndirectedGraphNode(n.label)
                    dic[n] = temp
                    dic[cur].neighbors += [temp]
                    queue.append(n)
        return newnode
        

File: cpp/101-150/133. Clone Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode *> neighbors;
    UndirectedGraphNode(int x) : label(x) {};
};
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (!node)
            return NULL;
        UndirectedGraphNode *copy = new UndirectedGraphNode(node->label);
        mp[node] = copy;
        queue<UndirectedGraphNode *> level;
        level.push(node);
        while (!level.empty()) {
        	UndirectedGraphNode *cur = level.front();
        	level.pop();
            for (auto neighbor : cur->neighbors) {
                if (!mp[neighbor]) {
                    UndirectedGraphNode *nb_copy = new UndirectedGraphNode(neighbor->label);
                    mp[neighbor] = nb_copy;
                    level.push(neighbor);
                }
                mp[cur]->neighbors.push_back(mp[neighbor]);
            }
        }
        return mp[node];
    }
private:
    unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> mp;
};